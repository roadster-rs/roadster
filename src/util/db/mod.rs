#[cfg(any(feature = "db-diesel-postgres", feature = "db-diesel-mysql"))]
pub(crate) mod diesel;
#[cfg(feature = "db-sea-orm")]
pub(crate) mod sea_orm;
