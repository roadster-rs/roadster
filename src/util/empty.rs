/// A placeholder that implements various traits so it can be used as the default for various type
/// parameters
pub struct Empty;

// Note: Unfortunately, this can't be implemented for any `impl App` because of a loop in the
// type resolution logic. So, just implement for the concrete `RoadsterApp` for now. If Rust's
// logic is ever updated to allow this type resolution loop, then we can add an `A: App` type
// parameter to implement for any `impl App`.
#[cfg(feature = "cli")]
#[async_trait::async_trait]
impl<S> crate::api::cli::RunCommand<crate::app::RoadsterApp<S, Empty>, S> for Empty
where
    S: Clone + Send + Sync + 'static,
    crate::app::context::AppContext: axum_core::extract::FromRef<S>,
{
    async fn run(
        &self,
        _prepared_app: &crate::app::PreparedApp<crate::app::RoadsterApp<S, Empty>, S>,
    ) -> crate::error::RoadsterResult<bool> {
        Ok(false)
    }
}

#[cfg(feature = "cli")]
impl clap::Args for Empty {
    fn augment_args(cmd: clap::Command) -> clap::Command {
        cmd
    }

    fn augment_args_for_update(cmd: clap::Command) -> clap::Command {
        cmd
    }
}

#[cfg(feature = "cli")]
impl clap::FromArgMatches for Empty {
    fn from_arg_matches(_matches: &clap::ArgMatches) -> Result<Self, clap::Error> {
        Ok(Empty)
    }

    fn update_from_arg_matches(&mut self, _matches: &clap::ArgMatches) -> Result<(), clap::Error> {
        Ok(())
    }
}

#[cfg(feature = "db-sea-orm")]
impl sea_orm_migration::MigratorTrait for Empty {
    fn migrations() -> Vec<Box<dyn sea_orm_migration::MigrationTrait>> {
        Default::default()
    }
}
