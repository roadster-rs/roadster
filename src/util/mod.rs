pub mod serde_util;
