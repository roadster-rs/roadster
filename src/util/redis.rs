use async_trait::async_trait;
use bb8::PooledConnection;
use sidekiq::redis_rs::ToRedisArgs;
use sidekiq::{RedisConnection, RedisConnectionManager, RedisError};

/// Trait to help with mocking responses from Redis.
#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub(crate) trait RedisCommands {
    async fn zrange(
        &mut self,
        key: String,
        lower: isize,
        upper: isize,
    ) -> Result<Vec<String>, RedisError>;

    async fn zrem<V>(&mut self, key: String, value: V) -> Result<usize, RedisError>
    where
        V: ToRedisArgs + Send + Sync + 'static;
}

#[async_trait]
impl RedisCommands for PooledConnection<'_, RedisConnectionManager> {
    async fn zrange(
        &mut self,
        key: String,
        lower: isize,
        upper: isize,
    ) -> Result<Vec<String>, RedisError> {
        RedisConnection::zrange(self, key, lower, upper).await
    }

    async fn zrem<V>(&mut self, key: String, value: V) -> Result<usize, RedisError>
    where
        V: ToRedisArgs + Send + Sync + 'static,
    {
        RedisConnection::zrem(self, key, value).await
    }
}
