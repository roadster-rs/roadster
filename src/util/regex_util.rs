pub const UUID_REGEX: &str = r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}";
