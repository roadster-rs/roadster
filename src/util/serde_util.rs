use std::fmt::Display;
use std::str::FromStr;

use serde::{de, Deserializer, Serializer};
use serde_derive::{Deserialize, Serialize};
use url::Url;

/// Custom deserializer to allow deserializing a string field as the given type type [T], as long as
/// the type implements [FromStr].
pub fn deserialize_from_str<'de, D, T>(deserializer: D) -> Result<T, D::Error>
where
    D: Deserializer<'de>,
    T: FromStr,
{
    let s: String = serde::Deserialize::deserialize(deserializer)?;
    match T::from_str(&s) {
        Ok(v) => Ok(v),
        Err(_) => Err(de::Error::custom(
            "String could not be parsed as the desired type.",
        )),
    }
}

/// Custom deserializer to allow serializing the given type [T] as a string, as long as the type
/// implements [FromStr].
pub fn serialize_to_str<S, T>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    T: ToString,
{
    serializer.serialize_str(&value.to_string())
}

/// Type that can be used to deserialize a value to a URI or, if the value is not a valid URI, fall
/// back to deserializing as a string.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(untagged)]
pub enum UriOrString {
    Uri(Url),
    String(String),
}

impl Display for UriOrString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            UriOrString::Uri(inner) => write!(f, "{}", inner),
            UriOrString::String(inner) => write!(f, "{}", inner),
        }
    }
}
