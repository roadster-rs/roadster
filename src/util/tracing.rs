use tracing::{Value, field};

pub fn optional_trace_field<T>(value: Option<T>) -> Box<dyn Value>
where
    T: ToString,
{
    value
        .map(|x| Box::new(field::display(x.to_string())) as Box<dyn Value>)
        .unwrap_or(Box::new(field::Empty))
}
