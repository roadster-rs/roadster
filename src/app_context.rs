use crate::app::App;
use crate::config::app_config::AppConfig;
#[cfg(feature = "db-sql")]
use sea_orm::Database;
#[cfg(feature = "db-sql")]
use sea_orm::DatabaseConnection;
use std::sync::Arc;
#[cfg(feature = "sidekiq")]
use tracing::info;

#[derive(Clone)]
pub struct AppContext<T = ()> {
    inner: Arc<AppContextInner>,
    custom: Arc<T>,
}

impl<T> AppContext<T> {
    // This method isn't used when running tests; only the mocked version is used.
    #[cfg_attr(test, allow(dead_code))]
    // The `A` type parameter isn't used in some feature configurations
    #[allow(clippy::extra_unused_type_parameters)]
    pub(crate) async fn new<A>(config: AppConfig) -> anyhow::Result<AppContext<()>>
    where
        A: App,
    {
        #[cfg(feature = "db-sql")]
        let db = Database::connect(A::db_connection_options(&config)?).await?;

        #[cfg(feature = "sidekiq")]
        let (redis_enqueue, redis_fetch) = {
            let sidekiq_config = &config.service.sidekiq;
            let redis_config = &sidekiq_config.custom.redis;
            let redis = sidekiq::RedisConnectionManager::new(redis_config.uri.to_string())?;
            let redis_enqueue = {
                let pool = bb8::Pool::builder().min_idle(redis_config.enqueue_pool.min_idle);
                let pool = redis_config
                    .enqueue_pool
                    .max_connections
                    .iter()
                    .fold(pool, |pool, max_conns| pool.max_size(*max_conns));
                pool.build(redis.clone()).await?
            };
            let redis_fetch = if redis_config
                .fetch_pool
                .max_connections
                .iter()
                .any(|max_conns| *max_conns == 0)
            {
                info!(
                "Redis fetch pool configured with size of zero, will not start the Sidekiq processor"
            );
                None
            } else {
                let pool = bb8::Pool::builder().min_idle(redis_config.fetch_pool.min_idle);
                let pool = redis_config
                    .fetch_pool
                    .max_connections
                    .iter()
                    .fold(pool, |pool, max_conns| pool.max_size(*max_conns));
                Some(pool.build(redis.clone()).await?)
            };
            (redis_enqueue, redis_fetch)
        };

        let inner = AppContextInner {
            config,
            #[cfg(feature = "db-sql")]
            db,
            #[cfg(feature = "sidekiq")]
            redis_enqueue,
            #[cfg(feature = "sidekiq")]
            redis_fetch,
        };
        Ok(AppContext {
            inner: Arc::new(inner),
            custom: Arc::new(()),
        })
    }

    pub fn with_custom<NewT: 'static>(self, custom: NewT) -> AppContext<NewT> {
        AppContext {
            inner: self.inner,
            custom: Arc::new(custom),
        }
    }

    pub fn config(&self) -> &AppConfig {
        &self.inner.config
    }

    #[cfg(feature = "db-sql")]
    pub fn db(&self) -> &DatabaseConnection {
        &self.inner.db
    }

    #[cfg(feature = "sidekiq")]
    pub fn redis_enqueue(&self) -> &sidekiq::RedisPool {
        &self.inner.redis_enqueue
    }

    #[cfg(feature = "sidekiq")]
    pub fn redis_fetch(&self) -> &Option<sidekiq::RedisPool> {
        &self.inner.redis_fetch
    }

    pub fn custom(&self) -> &T {
        &self.custom
    }
}

#[derive(Debug)]
struct AppContextInner {
    config: AppConfig,
    #[cfg(feature = "db-sql")]
    db: DatabaseConnection,
    #[cfg(feature = "sidekiq")]
    redis_enqueue: sidekiq::RedisPool,
    /// The Redis connection pool used by [sidekiq::Processor] to fetch Sidekiq jobs from Redis.
    /// May be `None` if the [fetch_pool.max_connections][crate::config::service::worker::sidekiq::ConnectionPool]
    /// config is set to zero, in which case the [sidekiq::Processor] would also not be started.
    #[cfg(feature = "sidekiq")]
    redis_fetch: Option<sidekiq::RedisPool>,
}

#[cfg(test)]
mockall::mock! {
    pub AppContext<T: 'static = ()> {
        pub fn config(&self) -> &AppConfig;

        #[cfg(feature = "db-sql")]
        pub fn db(&self) -> &DatabaseConnection;

        #[cfg(feature = "sidekiq")]
        pub fn redis_enqueue(&self) -> &sidekiq::RedisPool;

        #[cfg(feature = "sidekiq")]
        pub fn redis_fetch(&self) -> &Option<sidekiq::RedisPool>;

        pub fn with_custom<NewT: 'static>(self, custom: NewT) -> MockAppContext<NewT>;
    }

    impl<T> Clone for AppContext<T> {
        fn clone(&self) -> Self;
    }
}
