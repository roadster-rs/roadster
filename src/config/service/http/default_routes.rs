use crate::app_context::AppContext;
use crate::util::serde_util;
use crate::util::serde_util::default_true;
use serde_derive::{Deserialize, Serialize};
use validator::Validate;
use validator::ValidationError;

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[validate(schema(function = "validate_default_routes"))]
pub struct DefaultRoutes {
    #[serde(default = "default_true")]
    pub default_enable: bool,

    #[serde(deserialize_with = "deserialize_ping", default = "default_ping")]
    pub ping: DefaultRouteConfig,

    #[serde(deserialize_with = "deserialize_health", default = "default_health")]
    pub health: DefaultRouteConfig,

    #[cfg(feature = "open-api")]
    #[serde(
        deserialize_with = "deserialize_api_schema",
        default = "default_api_schema"
    )]
    pub api_schema: DefaultRouteConfig,

    #[cfg(feature = "open-api")]
    #[serde(deserialize_with = "deserialize_scalar", default = "default_scalar")]
    pub scalar: DefaultRouteConfig,

    #[cfg(feature = "open-api")]
    #[serde(deserialize_with = "deserialize_redoc", default = "default_redoc")]
    pub redoc: DefaultRouteConfig,
}

impl Default for DefaultRoutes {
    fn default() -> Self {
        Self {
            default_enable: default_true(),
            ping: default_ping(),
            health: default_health(),
            #[cfg(feature = "open-api")]
            api_schema: default_api_schema(),
            #[cfg(feature = "open-api")]
            scalar: default_scalar(),
            #[cfg(feature = "open-api")]
            redoc: default_redoc(),
        }
    }
}

fn validate_default_routes(
    // This parameter isn't used for some feature flag combinations
    #[allow(unused)] default_routes: &DefaultRoutes,
) -> Result<(), ValidationError> {
    #[cfg(feature = "open-api")]
    {
        let default_enable = default_routes.default_enable;
        let api_schema_enabled = default_routes.api_schema.enable.unwrap_or(default_enable);
        let scalar_enabled = default_routes.scalar.enable.unwrap_or(default_enable);
        let redoc_enabled = default_routes.redoc.enable.unwrap_or(default_enable);

        if scalar_enabled && !api_schema_enabled {
            return Err(ValidationError::new(
                "The Open API schema route must be enabled in order to use the Scalar docs route.",
            ));
        }
        if redoc_enabled && !api_schema_enabled {
            return Err(ValidationError::new(
                "The Open API schema route must be enabled in order to use the Redoc docs route.",
            ));
        }
    }

    Ok(())
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct DefaultRouteConfig {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub enable: Option<bool>,
    pub route: String,
}

impl DefaultRouteConfig {
    pub fn enabled<S>(&self, context: &AppContext<S>) -> bool {
        self.enable.unwrap_or(
            context
                .config()
                .service
                .http
                .custom
                .default_routes
                .default_enable,
        )
    }
}

// This fun boilerplate allows the user to
// 1. Partially override a config without needing to provide all of the required values for the config
// 2. Prevent a type's `Default` implementation from being used and overriding the default we
//    actually want. For example, we provide a default for the `route` fields, and we want that
//    value to be used if the user doesn't provide one, not the type's default (`""` in this case).
//
// See: https://users.rust-lang.org/t/serde-default-value-for-struct-field-depending-on-parent/73452/2
//
// This is mainly needed because all of the default routes share a struct for their common configs,
// so we can't simply set a default on the field directly with a serde annotation.
// An alternative implementation could be to have different structs for each default route's common
// config instead of sharing a struct type. However, that would still require a lot of boilerplate.

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "kebab-case", default)]
struct PartialDefaultRouteConfig {
    pub enable: Option<bool>,
    pub route: Option<String>,
}

fn deserialize_ping<'de, D>(deserializer: D) -> Result<DefaultRouteConfig, D::Error>
where
    D: serde::Deserializer<'de>,
{
    serde::Deserialize::deserialize(deserializer).map(map_empty_config("_ping".to_string()))
}

fn default_ping() -> DefaultRouteConfig {
    deserialize_ping(serde_util::empty_json_object()).unwrap()
}

fn deserialize_health<'de, D>(deserializer: D) -> Result<DefaultRouteConfig, D::Error>
where
    D: serde::Deserializer<'de>,
{
    serde::Deserialize::deserialize(deserializer).map(map_empty_config("_health".to_string()))
}

fn default_health() -> DefaultRouteConfig {
    deserialize_health(serde_util::empty_json_object()).unwrap()
}

#[cfg(feature = "open-api")]
fn deserialize_api_schema<'de, D>(deserializer: D) -> Result<DefaultRouteConfig, D::Error>
where
    D: serde::Deserializer<'de>,
{
    serde::Deserialize::deserialize(deserializer)
        .map(map_empty_config("_docs/api.json".to_string()))
}

#[cfg(feature = "open-api")]
fn default_api_schema() -> DefaultRouteConfig {
    deserialize_api_schema(serde_util::empty_json_object()).unwrap()
}

#[cfg(feature = "open-api")]
fn deserialize_scalar<'de, D>(deserializer: D) -> Result<DefaultRouteConfig, D::Error>
where
    D: serde::Deserializer<'de>,
{
    serde::Deserialize::deserialize(deserializer).map(map_empty_config("_docs".to_string()))
}

#[cfg(feature = "open-api")]
fn default_scalar() -> DefaultRouteConfig {
    deserialize_scalar(serde_util::empty_json_object()).unwrap()
}

#[cfg(feature = "open-api")]
fn deserialize_redoc<'de, D>(deserializer: D) -> Result<DefaultRouteConfig, D::Error>
where
    D: serde::Deserializer<'de>,
{
    serde::Deserialize::deserialize(deserializer).map(map_empty_config("_docs/redoc".to_string()))
}

#[cfg(feature = "open-api")]
fn default_redoc() -> DefaultRouteConfig {
    deserialize_redoc(serde_util::empty_json_object()).unwrap()
}

fn map_empty_config(
    default_route: String,
) -> impl FnOnce(PartialDefaultRouteConfig) -> DefaultRouteConfig {
    move |PartialDefaultRouteConfig { enable, route }| DefaultRouteConfig {
        enable,
        route: route.unwrap_or(default_route),
    }
}

#[cfg(test)]
mod tests {
    use crate::config::service::http::*;
    use rstest::rstest;

    #[rstest]
    #[case(false, false)]
    #[case(true, false)]
    #[cfg(not(feature = "open-api"))]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn validate_default_routes(#[case] default_enable: bool, #[case] validation_error: bool) {
        // Arrange
        #[allow(clippy::field_reassign_with_default)]
        let config = {
            let mut config = DefaultRoutes::default();
            config.default_enable = default_enable;
            config
        };

        // Act
        let result = config.validate();

        // Assert
        assert_eq!(result.is_err(), validation_error);
    }

    #[rstest]
    #[case(false, None, None, None, false)]
    #[case(true, None, None, None, false)]
    #[case(false, None, Some(true), None, true)]
    #[case(false, None, None, Some(true), true)]
    #[case(false, Some(true), Some(true), None, false)]
    #[case(false, Some(true), None, Some(true), false)]
    #[cfg(feature = "open-api")]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn validate_default_routes(
        #[case] default_enable: bool,
        #[case] api_schema_enabled: Option<bool>,
        #[case] scalar_enabled: Option<bool>,
        #[case] redoc_enabled: Option<bool>,
        #[case] validation_error: bool,
    ) {
        // Arrange
        #[allow(clippy::field_reassign_with_default)]
        let config = {
            let mut config = DefaultRoutes::default();
            config.default_enable = default_enable;
            config.api_schema.enable = api_schema_enabled;
            config.scalar.enable = scalar_enabled;
            config.redoc.enable = redoc_enabled;
            config
        };

        // Act
        let result = config.validate();

        // Assert
        assert_eq!(result.is_err(), validation_error);
    }
}

// To simplify testing, these are only run when all of the config fields are available
#[cfg(all(test, feature = "open-api"))]
mod deserialize_tests {
    use super::*;
    use crate::util::test_util::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case("")]
    #[case(
        r#"
        default-enable = false
        [ping]
        enable = true
        [health]
        enable = true
        [api-schema]
        enable = true
        [scalar]
        enable = true
        [redoc]
        enable = true
        "#
    )]
    #[case(
        r#"
        default-enable = false
        [ping]
        enable = false
        [health]
        enable = false
        [api-schema]
        enable = false
        [scalar]
        enable = false
        [redoc]
        enable = false
        "#
    )]
    #[case(
        r#"
        default-enable = false
        [ping]
        route = "a"
        [health]
        route = "b"
        [api-schema]
        route = "c"
        [scalar]
        route = "d"
        [redoc]
        route = "e"
        "#
    )]
    #[case(
        r#"
        [ping]
        enable = true
        route = "a"
        [health]
        enable = true
        route = "b"
        [api-schema]
        enable = true
        route = "c"
        [scalar]
        enable = true
        route = "d"
        [redoc]
        enable = true
        route = "e"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn auth(_case: TestCase, #[case] config: &str) {
        let default_routes: DefaultRoutes = toml::from_str(config).unwrap();

        assert_toml_snapshot!(default_routes);
    }
}
