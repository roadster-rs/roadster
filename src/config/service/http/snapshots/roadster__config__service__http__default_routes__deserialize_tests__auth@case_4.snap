---
source: src/config/service/http/default_routes.rs
expression: default_routes
---
default-enable = false

[ping]
route = 'a'

[health]
route = 'b'

[api-schema]
route = 'c'

[scalar]
route = 'd'

[redoc]
route = 'e'
